### PR TITLE
User registration and password reset are broken due to the last_authenticated value being null

### PR DESCRIPTION
**Pull request by srijeet0406**
This PR closes https://github.com/apache/trafficcontrol/issues/6457
<hr>

## Which Traffic Control components are affected by this PR?

- Traffic Ops
- Traffic Portal

## What is the best way to verify this PR?

Run TP and TO, and verify that you can successfully register new users and reset passwords of existing users.

## If this is a bugfix, which Traffic Control versions contained the bug?

- master

## PR submission checklist
- [x] This PR does not have tests 
- [x] This PR does not have documentation 
- [x] This PR has a CHANGELOG.md entry 
- [x] This PR **DOES NOT FIX A SERIOUS SECURITY VULNERABILITY** (see [the Apache Software Foundation's security guidelines](https://apache.org/security) for details)

### CODE REVIEW

File: traffic_ops/app/db/migrations/2022010715281600_remove_null_last_authenticated.up.sql
Comment (ocket8888) [inline traffic_ops/app/db/migrations/2022010715281600_remove_null_last_authenticated.up.sql:20]: Is this migration necessary now that the `last_authenticated` column is being properly updated on token/oauth login?

Comment (rawlinp) [reply traffic_ops/app/db/migrations/2022010715281600_remove_null_last_authenticated.up.sql:20]: It doesn't hurt really -- I think we could even make this column non-nullable with a default of beginning of epoch. There'd be one less nullable column to worry about at least.

Comment (ocket8888) [reply traffic_ops/app/db/migrations/2022010715281600_remove_null_last_authenticated.up.sql:20]: This is one thing that I think ought to be "nullable", because if a user has never logged in `null` is correct but the epoch value is a lie. Yes, it is impossible for a user to have actually logged in before ATC existed, but it's just nice to never lie, so that users/clients don't need to translate our responses themselves into what they actually probably mean - our system should be doing that for them, IMO.

I won't hold up the PR for it, but if we're gonna make it non-nullable the Go structures should use `time.Time` instead of `*time.Time` to reflect that.

Comment (srijeet0406) [reply traffic_ops/app/db/migrations/2022010715281600_remove_null_last_authenticated.up.sql:20]: I had it in there just as an added layer of protection, as Rawlin mentioned. I do agree with Brennan that if it is changed to non- nullable, we should change the Go structures as well. I'll let you guys decide which way we want to go.

Comment (ocket8888) [reply traffic_ops/app/db/migrations/2022010715281600_remove_null_last_authenticated.up.sql:20]: well, I think you have to be the tiebreaker here. I vote we let it be null and it looks to me like Rawlin is voting we don't, at the risk of speaking for him.

File: traffic_portal/app/src/common/api/UserService.js
Comment (ocket8888) [inline traffic_portal/app/src/common/api/UserService.js:105]: To keep things simple, I'd recommend just adding a new method e.g. `updateCurrentUser` and have the controller determine which method to call, the same way we have `getUsers` as well as `getCurrentUser`.

File: traffic_ops/app/db/migrations/2022010715281600_remove_null_last_authenticated.up.sql
Comment (rawlinp) [reply traffic_ops/app/db/migrations/2022010715281600_remove_null_last_authenticated.up.sql:20]: A null last_authenticated is still technically a lie for users that have authenticated before the feature was added, but I don't really care. We can't really answer the question of who has ever authenticated or not, only who has authenticated since this feature was deployed. If @ocket8888 prefers to keep the null values and delete these migrations, I think that's fine. I just generally like the idea of reducing the amount of nullable columns is all.

Comment (ocket8888) [reply traffic_ops/app/db/migrations/2022010715281600_remove_null_last_authenticated.up.sql:20]: For the record, I also like that idea, in general, just not in this case. It does feel a little strange arguing _in favor_ of nullability.

We could improve it a bit - without, still, being able to always tell the complete truth in all situations - by adding a migration that sets `last_authenticated = now() WHERE last_authenticated IS NULL` with no "down" statements and no `NOT NULL` constraint/nullability change. That way you know that all users haven't logged in since _at least_ their `lastAuthenticated` time, which is set to the date/time at which the feature was added (roughly - it's been at least one minor release since then in reality, but it's behind an unstable API atm). I'd find that more acceptable than setting it to Jan 1, 1970. That way there's no synonym for `null` being used and `null` is still allowed for new users moving forward.

File: traffic_ops/traffic_ops_golang/login/login.go
Comment (ocket8888) [inline traffic_ops/traffic_ops_golang/login/login.go:381]: The "Last Authenticated" time of a user should only be updated if the user successfully authenticates, which isn't checked until line 388 currently.

File: traffic_portal/app/src/modules/private/user/UserController.js
Comment (ocket8888) [inline traffic_portal/app/src/modules/private/user/UserController.js:54]: comparisons should use `===` - under what conditions was this occurring?

File: traffic_ops/traffic_ops_golang/login/login.go
Comment (ocket8888) [reply traffic_ops/traffic_ops_golang/login/login.go:237]: Actually, disregard, because it looks like you're double-logging this, because `api.HandleErr` will log it too. It looks like what you actually want something like:
```go
if dbErr != nil {
    dbErr = fmt.Errorf("unable to update authentication time for user '%s': %w", username, dbErr)
    api.HandleErr(w, r, nil, http.StatusInternalServerError, nil, dbErr)
    return
}
```
(I think the username is helpful information)

File: traffic_ops/app/db/migrations/2022010715281600_remove_null_last_authenticated.up.sql
Comment (srijeet0406) [reply traffic_ops/app/db/migrations/2022010715281600_remove_null_last_authenticated.up.sql:20]: Sounds good, I'll set it to `now()` wherever it is `null`. Fun fact, I had it as `now()` when I first started debugging this issue, and then changed it to epoch start later.

File: traffic_portal/app/src/modules/private/user/UserController.js
Comment (srijeet0406) [reply traffic_portal/app/src/modules/private/user/UserController.js:54]: This happens when the user tries to Update the password after clicking on the `Reset Password` link in TP

File: traffic_portal/app/src/common/api/UserService.js
Comment (ocket8888) [inline traffic_portal/app/src/common/api/UserService.js:87]: `currUser` is never reassigned, could be `const`.

File: traffic_portal/app/src/common/api/UserService.js
Comment (ocket8888) [inline traffic_portal/app/src/common/api/UserService.js:88]: nit but `: user` is superfluous; `{ user }` is equivalent to `{ user: user }`. I find that to be cleaner, but if you're not used to it I bet you could be confused by that syntax. So change it or don't, just wanted to point it out.

Comment (srijeet0406) [reply traffic_portal/app/src/common/api/UserService.js:88]: Yeah, I'm not used to the JS syntax at all. But I'll change it 👍

**Review by ocket8888 — APPROVED** [review]
This appears to fix the issue it's trying to fix, but I've discovered something troubling: the user is not required to set a password. The form can be submitted without setting a password, and that makes the token get dropped from the database, leaving the user with no way to authenticate, and that's not fixable without direct database manipulation. Also their cookie is immediately invalid so that the very next request 401's and there's no way to obtain a new cookie.

To reproduce: register a user, then go to the link in the email, set the full name as indicated by the form as being required, then submit the form. If you just wait like a minute a "newcount" logs request will be made that will cause a "user not found" error to be returned from the API and you'll be dropped back on the login page with no way to log in since you don't have a password.

This should be fixed, but if you think it's out-of-scope for this PR/just don't want to do it, I'll just merge it and open a new issue.

**Comment by ocket8888** [comment]
Solving that problem is more complicated than I initially thought, because of LDAP and OAuth users. I'll open an issue for it probably tomorrow. This PR doesn't need to try to fix the problem.

**Pull request merged**
